### ISSUE TITLE
return featrue maps in wrong oder?

### ISSUE THREAD

**Issue by weiyunchen92** [maunzzz/fine-grained-segmentation-networks#2]
in your model folder for pspnet.py in line 194 
        if self.training:
            if self.output_all:
                aux = self.conv6_1relu(aux_feat)
                aux = self.conv6_1(aux)
                h = self.conv6relu(h_feat)
                h = self.conv6(h)
                return h_feat, aux_feat, h, aux
            elif self.output_features:
                return h_feat, aux_feat
            else:
                aux_feat = self.conv6_1relu(aux_feat)
                aux = self.conv6_1(aux_feat)
                h_feat = self.conv6relu(h_feat)
                h = self.conv6(h_feat)
                return aux, h------------this one, maybe return h, aux

when you are training ,you want to calculate the resnet loss and auxiliary loss from resnet and auxiliary layer. But in here you return in a wrong oder. At the end auxiliary loss will be the main loss for you loss function instead of resnet loss.
      outputs_ref, aux_ref = net(img_ref) ---- line409 in train_with_clustering.py
.....
      main_loss, _ = loss_fct(outputs_ref, outputs_other, None, pts_ref,
                                       pts_other, cluster_labels=cluster_labels) ---- line437

**Comment by weiyunchen92** [comment]
pspnet.py
`class PSPNetClustering(nn.Module):
   ......

    def __init__(self, n_classes=19, n_blocks=[3, 4, 23, 3], pyramids=[6, 3, 2, 1], input_size=[713, 713], use_bn=True, output_features=False, output_all=False, use_original_base=False):

        super(PSPNetClustering, self).__init__()
        self.input_size = input_size
        self.n_classes = n_classes
        self.output_features = output_features
        self.output_for_cluster = False
        self.output_all = output_all
        if use_original_base:
            self.fcn = ResNetForPsp(Bottleneck, n_blocks)
        else:
            self.fcn = _DilatedFCN(n_blocks=n_blocks, use_bn=use_bn)
        # pyramid+cnnfeaturemap
        self.ppm = _PyramidPoolModule(
            in_channels=2048, pyramids=pyramids, use_bn=use_bn)
        self.final = nn.Sequential(OrderedDict([
            ('conv5_4', _ConvBatchNormReLU(4096, 512,
                                           3, 1, 1, 1, relu=False, use_bn=use_bn)),
            ('drop5_4', nn.Dropout2d(p=0.1)),
        ]))
        self.conv6relu = nn.ReLU()
        self.conv6 = nn.Conv2d(512, n_classes, 1, stride=1, padding=0)
        self.aux = nn.Sequential(OrderedDict([
            ('conv4_aux', _ConvBatchNormReLU(
                1024, 256, 3, 1, 1, 1, relu=False, use_bn=use_bn)),
            ('drop4_aux', nn.Dropout2d(p=0.1)),
        ]))
        self.conv6_1relu = nn.ReLU()
        self.conv6_1 = nn.Conv2d(256, n_classes, 1, stride=1, padding=0)

    def forward(self, x):
        x_size = x.size()
        if self.training:
            aux, h = self.fcn(x)
            aux_feat = self.aux(aux)
        else:
            h = self.fcn(x)

        h = self.ppm(h)
        h_feat = self.final(h)

        if self.training:
            if self.output_all:
                aux = self.conv6_1relu(aux_feat)
                aux = self.conv6_1(aux)
                h = self.conv6relu(h_feat)
                h = self.conv6(h)
                return h_feat, aux_feat, h, aux
            elif self.output_features:
                return h_feat, aux_feat
            else:
                aux_feat = self.conv6_1relu(aux_feat)
                aux = self.conv6_1(aux_feat)
                h_feat = self.conv6relu(h_feat)
                h = self.conv6(h_feat)
                return aux, h----------------this one
        else:
            if self.output_all:
                h = self.conv6relu(h_feat)
                h = self.conv6(h)
                return h_feat, h

            elif self.output_features:
                return h_feat
            else:
                h_feat = self.conv6relu(h_feat)
                h = self.conv6(h_feat)
                return h`

train_with_clustering.py
`
                    if args['feature_hinge_loss_weight'] > 0:
                        out_feat_ref, aux_feat_ref, outputs_ref, aux_ref = net(img_ref)
                    else:
                        outputs_ref, aux_ref = net(img_ref) -------here you return in h,aux
                     
                    ------auxiliary loss will be main loss
                    seg_main_loss = seg_loss_fct(outputs_ref, cluster_image)
                    seg_aux_loss = seg_loss_fct(aux_ref, cluster_image)

                    with torch.no_grad():
                        if args['feature_hinge_loss_weight'] > 0:
                            out_feat_other, aux_feat_other, outputs_other, aux_other = net(img_other)
                        else:
                            outputs_other, aux_other = net(img_other)
                else:  # use target
                    with torch.no_grad():
                        if args['feature_hinge_loss_weight'] > 0:
                            out_feat_ref, aux_feat_ref, outputs_ref, aux_ref = net(img_ref)
                        else:
                            outputs_ref, aux_ref = net(img_ref)

                    if args['feature_hinge_loss_weight'] > 0:
                        out_feat_other, aux_feat_other, outputs_other, aux_other = net(img_other)
                    else:
                        outputs_other, aux_other = net(img_other)

                    seg_main_loss = 0.
                    seg_aux_loss = 0.

                if args['feature_hinge_loss_weight'] > 0:
                    net.output_all = False

                ---------here will be auxiliary loss
                main_loss, _ = loss_fct(outputs_ref, outputs_other, None, pts_ref,
                                        pts_other, cluster_labels=cluster_labels) actually
                aux_loss, _ = loss_fct(aux_ref, aux_other, None, pts_ref, pts_other, cluster_labels=cluster_labels)

                if args['feature_hinge_loss_weight'] > 0:
                    feature_loss = feature_loss_fct(out_feat_ref, out_feat_other, pts_ref, pts_other, None)
                    feature_loss_aux = feature_loss_fct(aux_feat_ref, aux_feat_other, pts_ref, pts_other, None)
                    loss = args['corr_loss_weight']*(main_loss + 0.4 * aux_loss) + args['seg_loss_weight']*(seg_main_loss + 0.4 * seg_aux_loss) \
                        + args['feature_hinge_loss_weight']*(feature_loss + 0.4 * feature_loss_aux)
                else:
                    feature_loss = 0.
                    loss = args['corr_loss_weight']*(main_loss + 0.4 * aux_loss) + \
                        args['seg_loss_weight']*(seg_main_loss + 0.4 * seg_aux_loss)
`

**Comment by maunzzz** [comment]
You are correct! Pushed a fix,
Thanks!
